src/webcore/promise.rs

```diff
@@ -0,0 +1,260 @@
+            return Object( input ) === input &&
+                   typeof input.then === "function";
+        }).try_into().unwrap()
+    }
+
+    /// This function should rarely be needed, use [`PromiseFuture`](struct.PromiseFuture.html) instead.
+    ///
+    /// This function is needed if you have a JavaScript value which is a Promise-like object
+    /// (it has a `then` method) but it isn't a true `Promise`.
+    ///
+    /// That situation is rare, but it can happen if you are using a Promise library such as jQuery or
+    /// Bluebird.
+    ///
+    /// In that situation you can use `Promise::convert(value)` to convert it into a true `Promise`.
+    ///
+    /// If the `input` isn't a Promise-like object then it returns `None`.
+    ///
+    /// # Examples
+    ///
+    /// Convert a Promise-like object to a `Promise`:
+    ///
+    /// ```rust
+    /// // jQuery Promise
+    /// Promise::convert(js!( return $.get("test.php"); ))
+    ///
+    /// // Bluebird Promise
+    /// Promise::convert(js!( return bluebird_promise.timeout(1000); ))
+    /// ```
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve)
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-promise.resolve
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-promise-resolve-functions
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-promiseresolvethenablejob
+    pub fn convert( input: Value ) -> Option< Self > {
+        // TODO this can probably be made more efficient
+        if Promise::is_promise_like( &input ) {
+            Some( js!( return Promise.resolve( @{input} ); ).try_into().unwrap() )
+
+        } else {
+            None
+        }
+    }
+
+    /// This method is usually not needed, use [`PromiseFuture`](struct.PromiseFuture.html) instead.
+    ///
+    /// When the `Promise` either succeeds or fails, it calls the `callback` with the result.
+    ///
+    /// It does not wait for the `Promise` to succeed / fail (it does not block the thread).
+    ///
+    /// The `callback` is guaranteed to be called asynchronously even if the `Promise` is already succeeded / failed.
+    ///
+    /// If the `Promise` never succeeds / fails then the `callback` will never be called, and it will leak memory.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// promise.done(|result| {
+    ///     match result {
+    ///         Ok(success) => { ... },
+    ///         Err(error) => { ... },
+    ///     }
+    /// });
+    /// ```
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then)
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-performpromisethen
+    pub fn done< A, B >( &self, callback: B )
+        where A: TryFrom< Value >,
+              A::Error: std::error::Error,
+              B: FnOnce( Result< A, Error > ) + 'static {
```

Comment:
The error handling needs some work here.

1. I don't think using `Error` (as in - the JavaScript type `Error`) here is appropriate, as *technically* you can throw things which are not `Error` objects in JavaScript. It should be possible for the user to be able to retrieve the original value from promise's rejection.
2. It would be nice to figure out how to fit the `ConversionError` in this somehow. Perhaps convert it to a JavaScript `TypeError`? (This issue is also relevant in many other places across `stdweb` when we currently panic, but where I'd like to eventually throw a JavaScript exception when a type conversion fails.)

Reply:
Ah, good catch, you're right that technically any error can be thrown, so it should be `Result<Value, Value>` (which can be auto-coerced into something more precise by the user if they desire).

> It would be nice to figure out how to fit the ConversionError in this somehow.

What do you mean? It doesn't panic, it converts the `ConversionError` into a JavaScript `Error`. So what are you suggesting? To convert it into a `TypeError` instead?

Reply:
> What do you mean? It doesn't panic, it converts the ConversionError into a JavaScript Error. So what are you suggesting? To convert it into a JavaScript TypeError instead?

Yes, that's what I'm suggesting, since essentially a `ConversionError` *is* our counterpart of JS' `TypeError`. Another way to handle this would be to have an enum with two variants: one would be a `Value`, and another would be... something else (perhaps even the `ConversionError` if we make its innards private) which would signify a type error but without converting it to a JS type (keeping it purely in the Rust land).

I'm not sure which one would be better, but I'm *probably* leaning towards the first one's direction.

Reply:
I tried changing it to allow for arbitrary types for the error, but I got a lot of type errors (that I wasn't able to fix), so for now I'm just going to `.unwrap()` the `ConversionError` errors, but I'll add a TODO note to fix it later.